src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -515,11 +497,67 @@ protected void OnError(ErrorEventArgs e)
         [SuppressMessage("Microsoft.Security", "CA2109:ReviewVisibleEventHandlers", MessageId = "0#", Justification = "Changing from protected to private would be a breaking change")]
         protected void OnRenamed(RenamedEventArgs e)
         {
-            RenamedEventHandler renamedHandler = _onRenamedHandler;
-            if (renamedHandler != null)
+            _onRenamedHandler?.Invoke(this, e);
+        }
+
+        public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType) => 
+            WaitForChanged(changeType, Timeout.Infinite);
+
+        public WaitForChangedResult WaitForChanged(WatcherChangeTypes changeType, int timeout)
+        {
+            // The full framework implementation doesn't do any argument validation, so
+            // none is done here, either.
+
+            // Create a TCS, used as a convenient way to communicate a result from
+            // one thread to another blocked thread.
+            var tcs = new TaskCompletionSource<WaitForChangedResult>();
+            
+            // Create the event handlers to use to complete the task when an appropriate event occurs.
+            FileSystemEventHandler fseh = (s, e) =>
+            {
+                if ((e.ChangeType & changeType) != 0)
+                    tcs.TrySetResult(new WaitForChangedResult(e.ChangeType, e.Name, null, false));
+            };
+            RenamedEventHandler reh = (s, e) =>
```

Comment:
If we move up the ```((changeType & WatcherChangeTypes.Renamed) != 0)``` check then we can avoid making this  RenamedEventHandler for most cases.